### ISSUE TITLE
[Feature request] Fused location variant for Android?

### ISSUE THREAD

**Issue by Cheesebaron** [jamesmontemagno/GeolocatorPlugin#30]
FusedLocationApi uses less power and much faster than GPS only. Would love a variant of the GeolocatorPlugin which uses the Fused location API

**Comment by jamesmontemagno** [comment]
Yeah, i would like to do this. I wanted to create a new package that would bring in the google play dependencies, but didn't want to require it. Perhaps a package that could sit on tip of it and replace the functionality? Thoughts? or maybe make all the Android calls virtual to override in the extension package?

**Comment by jamesmontemagno** [comment]
Ideally the thought is that if Google Play Services isn't installed then it would fall back to default implementation.

**Comment by ep2016** [comment]
James, I thought your plugin does already work with Fused API, meaning that when everything is open on Android (5.0 is my version), i.e. a) Data connection, b) Location(High Accuracy) and c) [WiFi or always scan WiFi], the plugin reacts to a location request quickly and accurately. Anyway, my implementation before meeting your plugin was the fall back approach you referred to on your last comment. The problem I faced with that approach was that I keep receiving inaccurate coordinates from network provider, while GPS takes ages to reply when using the "old" android location method with "High Accuracy" having requested location updates from both providers. If I can help further, I would be happy to do so.
My issue with your plugin is that when I do not have WiFi open and do not have activated the "always scan for WiFi connections even if WiFi is disabled" setting and have my Location setting at "Device only", GetPositionAsync at the PCL project leads to an Exception saying that "A Task was cancelled". Is there any suggestion on that? Isn't the plugin supposed to work with GPS only?

**Comment by MisterJimson** [comment]
I have completed a basic implementation here:  https://github.com/MisterJimson/Geolocator.FusedLocationProvider

It can be used on top of this plugin, just swap out the Android implementation.  I didn't make a pull request since not much development is happening here and the only branch does not build for me.

My implementation still needs work to properly take into account the parameters and timeouts.

If I start to see some activity here I'll make a PR.

**Comment by jamesmontemagno** [comment]
@MisterJimson Very cool! I will take a look into it, this plugin is on my list to be updated to finish my 4.0 push.

**Comment by jamesmontemagno** [comment]
@MisterJimson I would love to pull it in to make it an official extension onto the project ;)

**Comment by MisterJimson** [comment]
Sure.  I'll continue to improve it as I require this implementation today.  When 4.0 is out or has a branch on Github I will make it a PR.

**Comment by jamesmontemagno** [comment]
So, the master branch is 4.0 actually, it had some legacy from the old repo. Will try to get it cleaned up next week. We can create an official extension package that we both publish out, which would be rad! I will put them both in CI and deploy at the same time :)

**Comment by MisterJimson** [comment]
Ah I see. Sounds good.

**Comment by Ivan-L** [comment]
Hi there @jamesmontemagno - is there still a plan to bring in @MisterJimson's work into the plugin? It would be great to use the Fused Location if it is available on Android. Here's a +1 from me!

**Comment by jamesmontemagno** [comment]
Yes, absolutely. I really want to do it, just time commitment.

My plan is to basically get a 4.0 out which I think is done to be honest. Then do another package called Geolocator+ which will have it based on Fused Location with GPS stuff

**Comment by jamesmontemagno** [comment]
My new new implementation has tons of improvements. I still want to implement this somehow though in teh future

**Comment by Steve-Himself** [comment]
I've submitted a pull request #128 for the FusedGeolocator implementation based on top of 4.0.1+. I did not change the build process as yet as it should be published as a new nuget package Xam.Plugin.Geolocator.Fused.

Essentially it's based on @MisterJimson's fork with a few tweaks and fixes. I have also changed the CrossGeolocator singleton helper, adding a new Init method which takes an Func<IGeolocator>. This provides an extensibility point if developers need to have other implementations for unit/automation testing. Could be a good pattern to apply to other plugins also.
